ctapipe/tools/stage1.py

```diff
@@ -0,0 +1,731 @@
+            "but add some overhead to the file. They can also be generated "
+            "and attached after the file is written "
+        ),
+        default_value=False,
+    ).tag(config=True)
+
+    overwrite = Bool(help="overwrite output file if it exists").tag(config=True)
+    progress_bar = Bool(help="show progress bar during processing").tag(config=True)
+
+    aliases = {
+        "input": "EventSource.input_url",
+        "output": "Stage1ProcessorTool.output_path",
+        "allowed-tels": "EventSource.allowed_tels",
+        "max-events": "EventSource.max_events",
+        "image-extractor-type": "Stage1ProcessorTool.image_extractor_type",
+        "gain-selector-type": "Stage1ProcessorTool.gain_selector_type",
+        "image-cleaner-type": "Stage1ProcessorTool.image_cleaner_type",
+    }
+
+    flags = {
+        "write-images": (
+            {"Stage1ProcessorTool": {"write_images": True}},
+            "store DL1/Event/Telescope images in output",
+        ),
+        "write-parameters": (
+            {"Stage1ProcessorTool": {"write_parameters": True}},
+            "store DL1/Event/Telescope parameters in output",
+        ),
+        "write-index-tables": (
+            {"Stage1ProcessorTool": {"write_index_tables": True}},
+            "generate PyTables index tables for the parameter and image datasets",
+        ),
+        "overwrite": (
+            {"Stage1ProcessorTool": {"overwrite": True}},
+            "Overwrite output file if it exists",
+        ),
+        "progress": (
+            {"Stage1ProcessorTool": {"progress_bar": True}},
+            "show a progress bar during event processing",
+        ),
+    }
+
+    classes = List(
+        [EventSource, CameraCalibrator, ImageQualityQuery]
+        + classes_with_traits(ImageCleaner)
+        + classes_with_traits(ImageExtractor)
+        + classes_with_traits(GainSelector)
+    )
+
+    def setup(self):
+
+        # prepare output path:
+        self.output_path = self.output_path.expanduser()
+        if self.output_path.exists():
+            if self.overwrite:
+                self.log.warning(f"Overwriting {self.output_path}")
+                self.output_path.unlink()
+            else:
+                self.log.critical(
+                    f"Output file {self.output_path} exists"
+                    ", use `--overwrite` to overwrite "
+                )
+                sys.exit(1)
+
+        PROV.add_output_file(str(self.output_path), role="DL1/Event")
+
+        # check that options make sense:
+        if self.write_parameters is False and self.write_images is False:
+            raise ToolConfigurationError(
+                "The options 'write_parameters' and 'write_images' are "
+                "both set to False. No output will be generated in that case. "
+                "Please enable one or both of these options."
+            )
+
+        # setup components:
+
+        self.gain_selector = self.add_component(
```

Comment:
Why must this be explicitly created instead of being able to configure this for the event source?

Reply:
This is the same pattern we use for the CameraCalibrator - the user passes in their own class.  It could be moved to the event source, of course. I think it was only because the SimTelEventSource only recently had the R1 calibration steps added.  ANyhow, let's leave that for a future refactoring, it's not really relevant here.